### ISSUE TITLE
jobTemplate section needed in charts/gha-runner-scale-set/values.yaml

### ISSUE THREAD

**Issue by snps-xlwu** [actions/actions-runner-controller#2945]
### What would you like added?
current we have listener Template and template for listener pod and runner pod,  we hope to have a separated job Template section

*A clear and concise description of what you want to happen.*

by adding jobTemplate , we can define  spec.volumes ,and spec.containers.<container id>.volumesMounts. so job container have access to some user-defined PVC , this is important to us as we have a huge size of build/regression dependencies which already exists in our NFS environment ( maintenanced by IT ) , what we need is to read or run  these tools from job container. we definitely don't hope to copy them into the ephemeral workspace because it is too huge ( > 10T ) 

Note: Feature requests to integrate vendor specific cloud tools (e.g. `awscli`, `gcloud-sdk`, `azure-cli`) will likely be rejected as the Runner image aims to be vendor agnostic.

### Why is this needed?

as described above. current job container only have access to "work" volume ( and all this subfolders to be mounted to special path from workflow job.<job id>.container.volumes part. we need to access any PVC defined in current namespace from job container.

*A clear and concise description of any alternative solutions or features you've considered.*

### Additional context

I am working in DevOps team in a Silicon Valley EDA company and our company is planning to migrate to GitHub from P4, a scalable runner is badly required. anyway, this is a gap that prevent us to use ARC to run build/regression. 

*Add any other context or screenshots about the feature request here.*

**Comment by apc-jnytnai0613** [comment]
Hi @snps-xlwu 

This is a very interesting way to add a volume. Incidentally, is it my understanding that you want to add jobTemplate to dind mode and kubernetes mode?

**Comment by snps-xlwu** [comment]
basically I want to have this feature for Kubernetes mode (  in dind mode we usually do not provide a container image in workflow, however in DIND mode I also find a problem that the dind container not by default mount the certification map and make it into /etc/ssl/certs like runner worker container did , so docker login command will fail when we use a self-signed docker registry , but this is another problem )

for this ticket, I mainly hope to have a job_Template section where I can define arbitrary user-defined PVCs and mount paths to my job container   so that in my job container  I can access to any resources which are mounted easily instead of building everything into my image (in fact that is impossible as that could be huge and could be changed day to day. ) . we cannot assuming that user can complete their job merely based on source code and a common image like a CentOS7 image. we need to mount lots of 3rd party libs or tools which is not suitable to build into image (that could be a large number of all kinds of images and hard to maintenance).

so, in summary, first, we need a way to define Volumes for job container. second, we need a way to define the mount path for these Volumes, the later one could be defined in workflow job part using some format like:

```
jobs:
  job1:
    container:
       image: "my image"
       mount:  vol-name:/path/in/container
```

as this could be more flexible for end users to select which volumes to be mounted to which path using same runner setting.
 
currently ARC only allow the vol-name part to be a sub dir under _work.  I hope it could support any volume name we define in the new added  job_template section.

**Comment by apc-jnytnai0613** [comment]
Hi, @snps-xlwu 
Thank you for your reply. Is it not possible to meet your needs by incorporating PVC into the runner container? Or is it mandatory that I want to attach it to the PVC of the job container?

**Comment by snps-xlwu** [comment]
let's check this example of runner template:

```
template:
  spec:
    containers:
      - name: runner
        image: ghcr.io/actions/actions-runner:latest
        command: [ "/home/runner/run.sh" ]
        env:
          - name: ACTIONS_RUNNER_CONTAINER_HOOKS
            value: /home/runner/k8s/index.js
          - name: ACTIONS_RUNNER_POD_NAME
            valueFrom:
              fieldRef:
                fieldPath: metadata.name
          - name: ACTIONS_RUNNER_REQUIRE_JOB_CONTAINER
            value: "false"
        volumeMounts:
          - name: work
            mountPath: /home/runner/_work
          - name: depot
            mountPath: /depot
    volumes:
      - name: depot  # I added an extra volume here
        persistentVolumeClaim:
          claimName: depot-pvc
      - name: work
        ephemeral:
          volumeClaimTemplate:
            spec:
              accessModes: [ "ReadWriteMany" ]
              storageClassName: "rook-cephfs"
              resources:
                requests:
                  storage: 1Gi
```

here "depot-pvc" is a PVC I defined in same namespace and it is mounted into /depot in runner pod. all these works perfectly and /depot is available in runner pod. however, in job container, the extra added volume "depot" is missed if you run "kubectl get pod <podname> -o yaml".  and because the volume is missed in job pod there is no way to mount it.

if you don't like to add extra job template, an acceptable workaround is to add these user-defined volumes into the runner template and arc copy these volumes definition into workflow pod. and provide a method to let user define the mount rules,

like in workflow yaml:

```
jobs:
  job1:
    container:
       image: "my image"
       mount:  depot:/path/in/container
```

here depot is a volume name we added into the runner template. (current ARC treat depot as a sub-dir under _work directory)

**Comment by apc-jnytnai0613** [comment]
Hi @snps-xlwu 
Kubernetes mode is implemented using [actions/runner-container-hooks](https://github.com/actions/runner-container-hooks).
Since actions/runner-container-hooks creates the Job Pod, it is better to ask the actions/runner-container-hooks side if it is possible to copy the volume.

The following is an extract of the actions/runner-container-hooks index.js file used when Runner runs in Kubernetes mode.
```node.js
function createPod(jobContainer, services, registry, extension) {
    var _a;
    return __awaiter(this, void 0, void 0, function () {
        var containers, appPod, instanceLabel, _b, claimName, secret, secretReference, body;
        var _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    containers = [];
                    if (jobContainer) {
                        containers.push(jobContainer);
                    }
                    if (services === null || services === void 0 ? void 0 : services.length) {
                        containers.push.apply(containers, services);
                    }
                    appPod = new k8s.V1Pod();
                    appPod.apiVersion = 'v1';
                    appPod.kind = 'Pod';
                    appPod.metadata = new k8s.V1ObjectMeta();
                    appPod.metadata.name = (0, constants_1.getJobPodName)();
                    instanceLabel = new constants_1.RunnerInstanceLabel();
                    appPod.metadata.labels = (_c = {},
                        _c[instanceLabel.key] = instanceLabel.value,
                        _c);
                    appPod.metadata.annotations = {};
                    appPod.spec = new k8s.V1PodSpec();
                    appPod.spec.containers = containers;
                    appPod.spec.restartPolicy = 'Never';
                    _b = appPod.spec;
                    return [4 /*yield*/, getCurrentNodeName()];
                case 1:
                    _b.nodeName = _d.sent();
                    claimName = (0, constants_1.getVolumeClaimName)();
                    appPod.spec.volumes = [
                        {
                            name: 'work',
                            persistentVolumeClaim: { claimName: claimName }
                        }
                    ];
                    if (!registry) return [3 /*break*/, 3];
                    return [4 /*yield*/, createDockerSecret(registry)];
                case 2:
                    secret = _d.sent();
                    if (!((_a = secret === null || secret === void 0 ? void 0 : secret.metadata) === null || _a === void 0 ? void 0 : _a.name)) {
                        throw new Error("created secret does not have secret.metadata.name");
                    }
                    secretReference = new k8s.V1LocalObjectReference();
                    secretReference.name = secret.metadata.name;
                    appPod.spec.imagePullSecrets = [secretReference];
                    _d.label = 3;
                case 3:
                    if (extension === null || extension === void 0 ? void 0 : extension.metadata) {
                        (0, utils_1.mergeObjectMeta)(appPod, extension.metadata);
                    }
                    if (extension === null || extension === void 0 ? void 0 : extension.spec) {
                        (0, utils_1.mergePodSpecWithOptions)(appPod.spec, extension.spec);
                    }
                    return [4 /*yield*/, k8sApi.createNamespacedPod(namespace(), appPod)];
                case 4:
                    body = (_d.sent()).body;
                    return [2 /*return*/, body];
            }
        });
    });
}
```

**Comment by nikola-jokic** [comment]
Hey @snps-xlwu,

I don't think we should extend ARC to provide a job template. There is a nice write-up [here](https://github.com/actions/actions-runner-controller/issues/2992#issuecomment-1764855221) showing how you can extend the job pod definition. You can leverage that to extend the job pod template and essentially achieve what you are trying to do :relaxed: 

Please keep in mind, you need to use hook version `0.4.0` to have this functionality. I will close this issue now, but feel free to comment on it

**Comment by snps-xlwu** [comment]
how to check my hook version? I am using  enterprise-server@3.9

**Comment by nikola-jokic** [comment]
Which version of the runner are you using? If it is the latest (2.310.2), it has the latest hook version

**Comment by snps-xlwu** [comment]
helm install arc  --namespace "arc-systems"  --create-namespace  oci://ghcr.io/actions/actions-runner-controller-charts/gha-runner-scale-set-controller

this is what I use to install ARC, how to check runner version?

**Comment by nikola-jokic** [comment]
The runner version comes from a different chart
The `gha-runner-scale-set` chart contains the template in which you are specifying the runner image. There, you can see the runner version based on the image tag

**Comment by snps-xlwu** [comment]
containers:
    - name: runner
      image: ghcr.io/actions/actions-runner:latest

it use :latest tag , from the doc, it is 2.310.2 , so It should be ok I guess

**Comment by snps-xlwu** [comment]
It works perfectly, just leave my implement here for others to reference.

create pod template configMap:  pod-template.yml:
```
apiVersion: v1
kind: ConfigMap                                                                                                                                                                                            
metadata:
  name: pod-templates
data:
  default.yaml: |
    ---
    apiVersion: v1
    kind: PodTemplate
    metadata:
      name: runner-pod-template
      labels:
        app: runner-pod-template
    spec:
      serviceAccountName: actions-runner-service-account
      containers:
        - name: $job
          volumeMounts:
            - name: depot
              mountPath: /depot
            - name: depotbld
              mountPath: /depotbld
            - name: remote-platforms
              mountPath: /remote/platforms
            - name: usr-local
              mountPath: /usr/local
      volumes:
        - name: depot
          persistentVolumeClaim:
            claimName: depot-pvc
        - name: depotbld
          persistentVolumeClaim:
            claimName: depotbld-pvc
        - name: remote-platforms
          persistentVolumeClaim:
            claimName: remote-platforms-pvc
        - name: usr-local
          persistentVolumeClaim:
            claimName: usr-local-pvc

```

kubectl -n arc-runners apply -f  pod-template.yml

runner-pod.values.yaml:
```

spec:
    serviceAccountName: actions-runner-service-account
    initContainers:
    - name: kube-init
      image: ghcr.io/actions/actions-runner:latest
      command: ["/bin/sh", "-c"]
      args:
        - |
          sudo chown -R 1001:123 /home/runner/_work
      volumeMounts:
        - name: work
          mountPath: /home/runner/_work
    containers:
      - name: runner
        image: ghcr.io/actions/actions-runner:latest
        command: [ "/home/runner/run.sh" ]
        env:
          - name: ACTIONS_RUNNER_CONTAINER_HOOKS
            value: /home/runner/k8s/index.js
          - name: ACTIONS_RUNNER_POD_NAME
            valueFrom:
              fieldRef:
                fieldPath: metadata.name
          - name: ACTIONS_RUNNER_REQUIRE_JOB_CONTAINER
            value: "false"
          - name: ACTIONS_RUNNER_CONTAINER_HOOK_TEMPLATE
            value: /home/runner/pod-templates/default.yaml
        volumeMounts:
          - name: work
            mountPath: /home/runner/_work
          - name: pod-templates
            mountPath: /home/runner/pod-templates
            readOnly: true
    volumes:
      - name: pod-templates
        configMap:
          name: pod-templates
      - name: work
        ephemeral:
          volumeClaimTemplate:
            spec:
              accessModes: [ "ReadWriteMany" ]
              storageClassName: "rook-cephfs"
              resources:
                requests:
                  storage: 20Gi

```

NOTE: in workflow job log, Initialize Containers step. there would be a warning, nice to remove it if possible:

Run '/home/runner/k8s/index.js'
Warning: Skipping name override: name can't be overwritten
